cpp/order/Steinhardt.cc

```
@@ -90,15 +99,37 @@ void Steinhardt::compute(const freud::locality::NeighborList* nlist,
 void Steinhardt::baseCompute(const freud::locality::NeighborList* nlist,
                              const freud::locality::NeighborQuery* points, freud::locality::QueryArgs qargs)
 {
-    const auto normalizationfactor = float(4.0 * M_PI / m_num_ms);
+    std::vector<float> normalizationfactor(m_ls.size());
+    for (auto l_index = 0; l_index < m_ls.size(); ++l_index)
+    {
+        normalizationfactor[l_index] = float(4.0 * M_PI / m_num_ms[l_index]);
+    }
     // For consistency, this reset is done here regardless of whether the array
     // is populated in baseCompute or computeAve.
-    m_qlm_local.reset();
+    for (auto& qlm_local : m_qlm_local)
+    {
+        qlm_local.reset();
+    }
+
     freud::locality::loopOverNeighborsIterator(
         points, points->getPoints(), m_Np, qargs, nlist,
         [=](size_t i, const std::shared_ptr<freud::locality::NeighborPerPointIterator>& ppiter) {
             float total_weight(0);
             const vec3<float> ref((*points)[i]);
+            // Construct PointSPHEvaluator outside loop since the construction is costly.
+            auto max_l = std::max_element(m_ls.begin(), m_ls.end());
+            if (max_l == m_ls.end())
+            {
+                throw std::invalid_argument("Selected l's do not have a maximum.");
```

Is this code path reachable? I don't think this is a case we need to catch. We would be able to simplify the code if this was removed (make `max_l` `const auto` and dereference the pointer immediately).

---

Only if they pass an empty list, which we can check for elsewhere.

---

That should raise an error in the constructor. Can't believe I didn't think of the empty case, I stared at that for a while. 🙃